File: Components/ASP.NET/VB/ASPxGridViewDemos/App_Code/HomesModel.vb
Imports Microsoft.VisualBasic
Imports System.Collections
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Data
Imports System.Drawing
Imports System.IO
Imports System.Linq
Imports System.Web
Imports DevExpress.Web.Data
Imports DevExpress.Web.Internal

Public Class HomesDataProvider
	Private Const ImageWidth As Integer = 250
	Private Const ImageFileNameFormat As String = "{0}.jpg", DataFileVirtualPath As String = "~/App_Data/Homes.xml", PhotoDirVirtualPath As String = "~/Content/HomesPhoto"

	Private Shared ReadOnly createLocker As Object = New Object()
	Private Shared homes_Renamed As List(Of Home) = Nothing

	Private Shared Function MapPath(ByVal path As String) As String
		Return HttpContext.Current.Server.MapPath(path)
	End Function
	Private Shared ReadOnly Property DataFilePath() As String
		Get
			Return MapPath(DataFileVirtualPath)
		End Get
	End Property
	Private Shared ReadOnly Property PhotoDirPath() As String
		Get
			Return MapPath(PhotoDirVirtualPath)
		End Get
	End Property
	Private Shared Function PhotoUrl(ByVal id As Integer) As String
		Return Path.Combine(PhotoDirVirtualPath, String.Format(ImageFileNameFormat, id))
	End Function

	Public Shared ReadOnly Property Homes() As List(Of Home)
		Get
			If homes_Renamed Is Nothing Then
				homes_Renamed = CreateHomes()
			End If
			Return homes_Renamed
		End Get
	End Property
	Public Shared Function SelectHomes() As List(Of Home)
		Return Homes
	End Function

	Private Shared Function CreateHomes() As List(Of Home)
		SyncLock createLocker
			Dim dataSet = New DataSet()
			dataSet.ReadXml(DataFilePath)
			Return dataSet.Tables(0).Select().Select(Function(r) ConvertToHome(r)).ToList()
		End SyncLock
	End Function

	Private Shared Function ConvertToHome(ByVal row As DataRow) As Home
		Dim result = New Home With {.ID = row.Field(Of Integer)("ID"), .Address = row.Field(Of String)("Address"), .Baths = row.Field(Of Short)("Baths"), .Beds = row.Field(Of Short)("Beds"), .HouseSize = row.Field(Of Double)("HouseSize"), .LotSize = row.Field(Of Double)("LotSize"), .Price = row.Field(Of Decimal)("Price"), .YearBuilt = row.Field(Of String)("YearBuilt"), .Photo = row.Field(Of Byte())("Photo"), .Features = row.Field(Of String)("Features"), .Type = CType(row.Field(Of Integer)("Type"), HomeType), .Status = CType(row.Field(Of Integer)("Status"), HomeStatus)}
		result.PhotoUrl = GetPhotoUrl(result)
		Return result
	End Function

	Private Shared Function GetPhotoUrl(ByVal item As Home) As String
		Dim url = PhotoUrl(item.ID)
		Dim serverPath = MapPath(url)

		If (Not File.Exists(serverPath)) Then
			If (Not Directory.Exists(PhotoDirPath)) Then
				Directory.CreateDirectory(PhotoDirPath)
			End If
			SaveImage(item.Photo, serverPath)
		End If
		Return url
	End Function
	Private Shared Sub SaveImage(ByVal imageBytes() As Byte, ByVal path As String)
		Using stream = New MemoryStream(imageBytes)
		Using img = Image.FromStream(stream)
			Dim size As New Size(ImageWidth, CInt(Fix(img.Height)) * ImageWidth / img.Width)
			Using thumbnail As Image = New ImageThumbnailCreator(img).CreateImageThumbnail(size)
				ImageUtils.SaveImage(thumbnail, path)
			End Using
		End Using
		End Using
	End Sub
End Class

Public Enum HomeType
	<Description("Single Family Home")> _
	SingleFamilyHome
	<Description("Condo/Townhouse")> _
	Townhome
	<Description("Multi-Family Home")> _
	MultiFamilyHome
End Enum
Public Enum HomeStatus
	<Description("New Construction")> _
	NewConstruction
	<Description("Foreclosures")> _
	Foreclosures
	<Description("Recently Sold")> _
	RecentlySold
End Enum

Public Class Home
	Public Property ID() As Integer
	Public Property Address() As String
	Public Property Baths() As Short
	Public Property Beds() As Short
	Public Property HouseSize() As Double
	Public Property LotSize() As Double
	Public Property Price() As Decimal
	Public Property YearBuilt() As String
	Public Property Photo() As Byte()
	Public Property PhotoUrl() As String
	Public Property Features() As String
	Public Property Type() As HomeType
	Public Property Status() As HomeStatus
End Class

File: Components/ASP.NET/CS/ASPxCardViewDemos/App_Code/HomesModel.cs
using System.Collections;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.IO;
using System.Linq;
using System.Web;
using DevExpress.Web.Data;
using DevExpress.Web.Internal;

public class HomesDataProvider {
    const int ImageWidth = 250;
    const string 
        ImageFileNameFormat = "{0}.jpg",
        DataFileVirtualPath = "~/App_Data/Homes.xml",
        PhotoDirVirtualPath = "~/Content/HomesPhoto";

    static readonly object createLocker = new object();
    static List<Home> homes = null;

    static string MapPath(string path) { return HttpContext.Current.Server.MapPath(path); }
    static string DataFilePath { get { return MapPath(DataFileVirtualPath); } }
    static string PhotoDirPath { get { return MapPath(PhotoDirVirtualPath); } }
    static string PhotoUrl(int id) { return Path.Combine(PhotoDirVirtualPath, string.Format(ImageFileNameFormat, id)); }

    public static List<Home> Homes {
        get {
            if(homes == null)
                homes = CreateHomes();
            return homes;
        }
    }
    public static List<Home> SelectHomes() { return Homes; }

    static List<Home> CreateHomes() {
        lock(createLocker) {
            var dataSet = new DataSet();
            dataSet.ReadXml(DataFilePath);
            return dataSet.Tables[0].Select().Select(r => ConvertToHome(r)).ToList();
        }
    }

    static Home ConvertToHome(DataRow row) {
        var result = new Home {
            ID = row.Field<int>("ID"),
            Address = row.Field<string>("Address"),
            Baths = row.Field<short>("Baths"),
            Beds = row.Field<short>("Beds"),
            HouseSize = row.Field<double>("HouseSize"),
            LotSize = row.Field<double>("LotSize"),
            Price = row.Field<decimal>("Price"),
            YearBuilt = row.Field<string>("YearBuilt"),
            Photo = row.Field<byte[]>("Photo"),
            Features = row.Field<string>("Features"),
            Type = (HomeType)row.Field<int>("Type"),
            Status = (HomeStatus)row.Field<int>("Status")
        };
        result.PhotoUrl = GetPhotoUrl(result);
        return result;
    }
    
    static string GetPhotoUrl(Home item) {
        var url = PhotoUrl(item.ID);
        var serverPath = MapPath(url);

        if(!File.Exists(serverPath)) {
            if(!Directory.Exists(PhotoDirPath))
                Directory.CreateDirectory(PhotoDirPath);
            SaveImage(item.Photo, serverPath);
        }
        return url;
    }
    static void SaveImage(byte[] imageBytes, string path) {
        using(var stream = new MemoryStream(imageBytes))
        using(var img = Image.FromStream(stream)) {
            Size size = new Size(ImageWidth, (int)img.Height * ImageWidth / img.Width);
            using (Image thumbnail = new ImageThumbnailCreator(img).CreateImageThumbnail(size))
                ImageUtils.SaveImage(thumbnail, path);
        }
    }
}

public enum HomeType {
    [Description("Single Family Home")]
    SingleFamilyHome,
    [Description("Condo/Townhouse")]
    Townhome,
    [Description("Multi-Family Home")]
    MultiFamilyHome
};
public enum HomeStatus {
    [Description("New Construction")]
    NewConstruction,
    [Description("Foreclosures")]
    Foreclosures,
    [Description("Recently Sold")]
    RecentlySold
};

public class Home {
    public int ID { get; set; }
    public string Address { get; set; }
    public short Baths { get; set; }
    public short Beds { get; set; }
    public double HouseSize { get; set; }
    public double LotSize { get; set; }
    public decimal Price { get; set; }
    public string YearBuilt { get; set; }
    public byte[] Photo { get; set; }
    public string PhotoUrl { get; set; }
    public string Features { get; set; }
    public HomeType Type { get; set; }
    public HomeStatus Status { get; set; }
}
